Fetch ram/rom start/size

aashishc1988 opened the pull request:
### Description

Fetch rom/ram start/size information and create defines for the same.

### Pull request type

    [x] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

aashishc1988 commented:
@theotherjimmy @deepikabhavnani

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -610,24 +610,71 @@ def _get_mem_specs(self, memories, cmsis_part, exception_text):
+                memory.split(),
+                cmsis_part,
+                "Not enough information in CMSIS packs to build a bootloader "
+                "project"
```
Could you indent this line and the 3 prion lines so that we know they are arguments to the `self._get_mem_specs` call?

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -610,24 +610,71 @@ def _get_mem_specs(self, memories, cmsis_part, exception_text):
+                )
+                rom_start = int(rom_start, 0)
+                rom_size = int(rom_size, 0)
+                if all_regions == False:
```
Could you use the more-pythonic:
```python
if not all_regions
```

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -610,24 +610,71 @@ def _get_mem_specs(self, memories, cmsis_part, exception_text):
+                rom_start = int(rom_start, 0)
+                rom_size = int(rom_size, 0)
+                if all_regions == False:
+                    return rom_start, rom_size
```
I'm not a fan of having different return values based on an input parameter. You are welcome to drop the `all_regions` parameter in favor of using `self.rom`.

aashishc1988 replied:
As discussed, we got rid of rom property, so we need input parameter and different return value based on that. let me know if you disagree and we can discuss other possible alternatives.

bridadan replied:
I'd agree with @theotherjimmy here on not changing the return value based on input if at all possible

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -610,24 +610,71 @@ def _get_mem_specs(self, memories, cmsis_part, exception_text):
+                    memory = 'ROM'
+                else:
+                    rom_active_memories += 1
+                    memory = 'ROM'+str(rom_active_memories)
```
Please either use string formatting:
```suggestion
                    memory = 'ROM{}'.format(str(rom_active_memories))
```

or add a space around the `+`.

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -610,24 +610,71 @@ def _get_mem_specs(self, memories, cmsis_part, exception_text):
+                memory.split(),
+                cmsis_part,
+                "Not enough information in CMSIS packs to build a bootloader "
+                "project"
```
Same as above, please indent this so that we know it's part of the `_get_mem_specs` call.

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -610,24 +610,71 @@ def _get_mem_specs(self, memories, cmsis_part, exception_text):
+        ram_regions = {}
+        # Counter to keep track of RAM memories supported by target
+        ram_active_memories = 0
+        ram_memories = ['IRAM1', 'IRAM2', 'IRAM3', 'IRAM4', 'SRAM_OC', 'SRAM_ITC', 'SRAM_DTC', 'SRAM_UPPER', 'SRAM_LOWER']
```
This line is really long. Please split it near 80 chars.

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -610,24 +610,71 @@ def _get_mem_specs(self, memories, cmsis_part, exception_text):
+                )
+                ram_start = int(ram_start, 0)
+                ram_size = int(ram_size, 0)
+                if (memory == 'IRAM1' or memory == 'SRAM_OC' or memory == 'SRAM_UPPER') and (not self.has_ram_regions):
```
This line is also pretty long, please split it near 80 chars.

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -610,24 +610,71 @@ def _get_mem_specs(self, memories, cmsis_part, exception_text):
+                    memory = 'RAM'
+                else:
+                    ram_active_memories += 1
+                    memory = 'RAM'+str(ram_active_memories)
```
Please either use string formatting:
```suggestion
                    memory = 'RAM{}'.format(str(rom_active_memories))
```

or add a space around the `+`.

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -610,24 +610,71 @@ def _get_mem_specs(self, memories, cmsis_part, exception_text):
+        return rom_regions
+
+    def _get_all_ram_regions(self):
+        """Get information of all available ram regions as a pair of [Region, [start_addr, size]]"""
```
Please reword this docstring to mention that this is a mapping/dictionary, and not a list of lists of lists.

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -653,13 +700,13 @@ def regions(self):
                 "target.mbed_app_size")
         if self.target.bootloader_img or self.target.restrict_size:
-            return self._generate_bootloader_build(*self.rom)
+            return self._generate_bootloader_build(*self._get_all_rom_regions())
```
I'm a little surprised by this change. What does it do? Why is it needed?

aashishc1988 replied:
It expects rom_start, rom_size as the output and that is then used. We got rid of rom property, so thats why this call to get those parameters.

theotherjimmy commented on tools/toolchains/__init__.py:
```diff
@@ -733,12 +733,27 @@ def add_regions(self):
-                suffixes=["_START", "_SIZE"]
-            )
+            # Add all available ROM regions to build profile
+            rom_available_regions = self.config._get_all_rom_regions(True)
```
This is not allowed: `_get_all_rom_regions` is private.

theotherjimmy commented on tools/toolchains/__init__.py:
```diff
@@ -733,12 +733,27 @@ def add_regions(self):
+                region, rom_start, rom_size = key, value[0], value[1]
+                Region = namedtuple("Region", "name start size")
+                self._add_defines_from_region(
+                    Region("MBED_"+region, rom_start, rom_size),
```
Please either add a space, or use string formatting here too.

theotherjimmy commented on tools/toolchains/__init__.py:
```diff
@@ -733,12 +733,27 @@ def add_regions(self):
+            pass
+        try:
+            # Add all available RAM regions to build profile
+            ram_available_regions = self.config._get_all_ram_regions()
```
This is not allowed: `_get_all_rom_regions` is private.

theotherjimmy commented on tools/toolchains/__init__.py:
```diff
@@ -733,12 +733,27 @@ def add_regions(self):
+        try:
+            # Add all available RAM regions to build profile
+            ram_available_regions = self.config._get_all_ram_regions()
+            for key,value in ram_available_regions.items():
```
Please add a ` ` (space char) after the `,`

theotherjimmy commented on tools/toolchains/__init__.py:
```diff
@@ -733,12 +733,27 @@ def add_regions(self):
+            # Add all available ROM regions to build profile
+            rom_available_regions = self.config._get_all_rom_regions(True)
+            for key,value in rom_available_regions.items():
+                region, rom_start, rom_size = key, value[0], value[1]
```
I have several issues with this line:
 * It creates an alias for key
 * It assigns `rom_start` and `rom_size` without destructuring
 * It uses magic numbers (the array indexes)

```suggestion
                rom_start, rom_size = value
```

theotherjimmy commented on tools/toolchains/__init__.py:
```diff
@@ -733,12 +733,27 @@ def add_regions(self):
+            # Add all available RAM regions to build profile
+            ram_available_regions = self.config._get_all_ram_regions()
+            for key,value in ram_available_regions.items():
+                region, ram_start, ram_size = key, value[0], value[1]
```
I have several issues with this line:
 * It creates an alias for key
 * It assigns `rom_start` and `rom_size` without destructuring
 * It uses magic numbers (the array indexes)

```suggestion
                ram_start, ram_size = value
```

theotherjimmy reviewed (changes requested):
Mostly formatting. A little use of private APIs outside their own class.

aashishc1988 commented:
Changes added. @theotherjimmy @deepikabhavnani please review

bridadan commented on tools/config/__init__.py:
```diff
@@ -610,24 +610,75 @@ def _get_mem_specs(self, memories, cmsis_part, exception_text):
+        for memory in rom_memories:
+            try:
+                rom_start, rom_size = self._get_mem_specs(
+                    memory.split(),
```
Is the point here just to form an array from a string? Typically `split` is used to deal with whitespace or delimited strings.

Would you be opposed to switching this to:
```suggestion
                    [memory],
```

We wouldn't expect `rom_memories` items to have white space correct? I'd prefer this to error if there was just hanging trailing whitespace in that array.

bridadan commented on tools/config/__init__.py:
```diff
@@ -610,24 +610,75 @@ def _get_mem_specs(self, memories, cmsis_part, exception_text):
+        rom_regions = {}
+        # Counter to keep track of ROM memories supported by target
+        rom_active_memories = 0
+        rom_memories = ['IROM1', 'PROGRAM_FLASH', 'IROM2']
```
Is this list coming from CMSIS packs?

bridadan commented on tools/config/__init__.py:
```diff
@@ -610,24 +610,75 @@ def _get_mem_specs(self, memories, cmsis_part, exception_text):
+        # Counter to keep track of RAM memories supported by target
+        ram_active_memories = 0
+        ram_memories = ['IRAM1', 'IRAM2', 'IRAM3', 'IRAM4', 'SRAM_OC', \
+                        'SRAM_ITC', 'SRAM_DTC', 'SRAM_UPPER', 'SRAM_LOWER']
```
Is this list coming from CMSIS packs?

bridadan commented on tools/config/__init__.py:
```diff
@@ -610,24 +610,75 @@ def _get_mem_specs(self, memories, cmsis_part, exception_text):
+        for memory in ram_memories:
+            try:
+                ram_start, ram_size = self._get_mem_specs(
+                    memory.split(),
```
Same question about `.split()` here. If it's just to turn a string into an array I'd recommend doing that explicitly:

```suggestion
                    [memory],
```

bridadan commented on tools/config/__init__.py:
```diff
@@ -610,24 +610,75 @@ def _get_mem_specs(self, memories, cmsis_part, exception_text):
+                    memory = 'RAM' + str(ram_active_memories)
+                ram_regions[memory] = [ram_start, ram_size]
+            except ConfigException:
+                continue         
```
(Sorry if I'm missing background info on this stuff, just coming up to speed on this stuff).

Are most of the regions specified in `ram_memories` expected to be present? If not, it seems a bit funny to loop through all the possible memory regions and just catching the exception. Looking at this code here: 

https://github.com/ARMmbed/mbed-os/blob/abdfcc0579f432b3c8450463b15cc7c5b4d4e7db/tools/config/__init__.py#L606-L607

Looks like it's just checking for keys in the dictionary `cmsis_part`. You could cut down a bit on the looping by doing a bit of set logic on the keys:

```python
ram_memories = set(['IRAM1', 'IRAM2', 'IRAM3', 'IRAM4', 'SRAM_OC', \
                             'SRAM_ITC', 'SRAM_DTC', 'SRAM_UPPER', 'SRAM_LOWER'])
present_memories = set(cmsis_part['memory'].keys())
valid_memories = ram_memories.intersect(present_memories)
for memory in valid_memories:
    # resume existing code but without the need for the try/except
```

aashishc1988 replied:
Great suggestion! I should have paid more attention to this. Thanks :)

bridadan commented on tools/config/__init__.py:
```diff
@@ -653,13 +704,13 @@ def regions(self):
              (self.target.mbed_app_start or self.target.mbed_app_size)):
             raise ConfigException(
-                "target.bootloader_img and target.restirct_size are "
+                "target.bootloader_img and target.restrict_size are "
```
Great catch 👍

bridadan commented on tools/toolchains/__init__.py:
```diff
@@ -733,12 +733,27 @@ def add_regions(self):
+            ram_available_regions = self.config.get_all_ram_regions()
+            for key, value in ram_available_regions.items():
+                ram_start, ram_size = value
+                Region = namedtuple("Region", "name start size")
```
I'd recommend removing this if you also go with the commonly declared `Region` I suggested above.

bridadan commented on tools/toolchains/__init__.py:
```diff
@@ -733,12 +733,27 @@ def add_regions(self):
+            rom_available_regions = self.config.get_all_rom_regions(True)
+            for key, value in rom_available_regions.items():
+                rom_start, rom_size = value
+                Region = namedtuple("Region", "name start size")
```
Would you mind moving this declaration out of the loop? (it'd probably be a good idea to move it out of the try/except block as well)

bridadan reviewed (changes requested):
Quite a few questions, a few suggested changes, but nothing scary 😄 good work!

aashishc1988 commented:
@bridadan @theotherjimmy Made some more changes. Can you guys review again?

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -610,24 +610,76 @@ def _get_mem_specs(self, memories, cmsis_part, exception_text):
-    @property
-    def rom(self):
-        """Get rom information as a pair of start_addr, size"""
+    def get_all_rom_regions(self):
```
As we are now creating 2 conflicting definitions for "Regions" I recomend using the term "Memories" for these functions.

Memories, by this definition, will be any hardware defined (via cmsis-packs and targets.json overrides) places to store stuff.
Regions, by this definition, will be a partition within the available memories.
The active regions are the regions that will be used during this link step.

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -610,24 +610,76 @@ def _get_mem_specs(self, memories, cmsis_part, exception_text):
+            
+        return rom_regions
+
+    def get_all_ram_regions(self):
```
Same as above: let's use "memories" in place of "regions" here

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -653,13 +705,13 @@ def regions(self):
                 "target.mbed_app_size")
         if self.target.bootloader_img or self.target.restrict_size:
-            return self._generate_bootloader_build(*self.rom)
+            return self._generate_bootloader_build(*self.get_all_rom_regions())
```
I'm pretty sure you meant:
```suggestion
            return self._generate_bootloader_build(**self.get_all_rom_regions())
```

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -653,13 +705,13 @@ def regions(self):
+            return self._generate_bootloader_build(*self.get_all_rom_regions())
         else:
-            return self._generate_linker_overrides(*self.rom)
+            return self._generate_linker_overrides(*self.get_all_rom_regions())
```
```suggestion
            return self._generate_linker_overrides(**self.get_all_rom_regions())
```

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -696,7 +748,8 @@ def _assign_new_offset(rom_start, start, new_offset, region_name):
         return newstart
 
-    def _generate_bootloader_build(self, rom_start, rom_size):
+    def _generate_bootloader_build(self, active_rom_regions):
```
This will have to become:
```suggestion
    def _generate_bootloader_build(self, ROM=None, **other_rom_memories):
```

aashishc1988 replied:
It will become, 
 `def _generate_bootloader_build(self, **other_rom_memories):` , right? Why `ROM=None`. I am only passing the dictionary.

theotherjimmy replied:
So that you can use `ROM` in the next diff.

aashishc1988 replied:
May be I am misunderstanding, or missing something. We do not have ROM property anymore, and given the fact that each target will have IROM1 or program flash defined, we will be passing the dictionary with ROM defined for sure. So for the next diff, instead of
`rom_start, rom_size = next(iter(active_rom_regions.values()))`
we can do
`rom_start, rom_size = active_rom_regions.get('ROM')`
It's very well possible that I am missing some high level view of the intention behind this function. Is it? Let me commit my latest changes and lets get back to this if needed.

theotherjimmy replied:
> We do not have ROM property anymore

That's unrelated.

> given the fact that each target will have IROM1 or program flash defined, we will be passing the dictionary with ROM defined for sure

This make the dictionary item `ROM` a local.

Nothing "high level" or anything like that. Feel free to implement it the way you're suggesting, with the modification `rom_start, rom_size = active_rom_regions['ROM']`.

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -696,7 +748,8 @@ def _assign_new_offset(rom_start, start, new_offset, region_name):
 
-    def _generate_bootloader_build(self, rom_start, rom_size):
+    def _generate_bootloader_build(self, active_rom_regions):
+        rom_start, rom_size = next(iter(active_rom_regions.values()))
```
In tandem with the above:
```suggestion
        rom_start, rom_size = ROM or next(iter(other_rom_memories.values()))
```

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -780,7 +833,8 @@ def report(self):
 
-    def _generate_linker_overrides(self, rom_start, rom_size):
+    def _generate_linker_overrides(self, active_rom_regions):
+        rom_start, rom_size = next(iter(active_rom_regions.values()))
```
Same diff as `_generate_bootloader_build` here.

theotherjimmy reviewed (changes requested):
This is looking better. Further comments below.

I think that `get_all_rom_regions` and `get_all_ram_regions` look very similar. Could we de-deplicate this into a single helper function that takes in the differences?

aashishc1988 commented:
Did some refactoring. Combined get_all_rom/ram regions function into one. Created a new function that gives the list of all available region for ram/rom, since this information is used at couple of places, it was better to put it one place than to duplicate.

aashishc1988 commented on tools/config/__init__.py:
```diff
@@ -610,24 +610,62 @@ def _get_mem_specs(self, memories, cmsis_part, exception_text):
+        else:
+            return ['IRAM1', 'IRAM2', 'IRAM3', 'IRAM4', 'SRAM_OC', \
+                    'SRAM_ITC', 'SRAM_DTC', 'SRAM_UPPER', 'SRAM_LOWER']
+
```
We could use get_cmsis part info and generate this table from it but that seems bit tricky (using substring to extract string) and that can get messy if new region names (which are slightly unique than usual) were added for future targets. if you can think of a better way, let me know. This seems slightly more maintainable.

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -610,24 +610,62 @@ def _get_mem_specs(self, memories, cmsis_part, exception_text):
-    @property
-    def rom(self):
-        """Get rom information as a pair of start_addr, size"""
+    def get_region_list(self, region='ROM'):
```
This function seems... unnecessary. It's a switch between 2 constants that could be defined at the top of the file.

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -610,24 +610,62 @@ def _get_mem_specs(self, memories, cmsis_part, exception_text):
+                    'SRAM_ITC', 'SRAM_DTC', 'SRAM_UPPER', 'SRAM_LOWER']
+
+    def get_all_active_memories(self, *memory_list):
+        """Get information of all available rom/ram regions in the form of dictionary 
```
Can you change the document comment to talk about memories instead of regions.

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -610,24 +610,62 @@ def _get_mem_specs(self, memories, cmsis_part, exception_text):
         # 1. Doesn't support CMSIS pack, or
         # 2. Supports TrustZone and user needs to change its flash partition
+        
+        available_regions = {}
```
Can we change all locals to reference memories instead of regions.

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -610,24 +610,62 @@ def _get_mem_specs(self, memories, cmsis_part, exception_text):
+            return ['IRAM1', 'IRAM2', 'IRAM3', 'IRAM4', 'SRAM_OC', \
+                    'SRAM_ITC', 'SRAM_DTC', 'SRAM_UPPER', 'SRAM_LOWER']
+
+    def get_all_active_memories(self, *memory_list):
```
You probably don't want to pass this in as all positional arguments. You should remove the `*`.

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -653,13 +691,13 @@ def regions(self):
                 "target.mbed_app_size")
         if self.target.bootloader_img or self.target.restrict_size:
-            return self._generate_bootloader_build(*self.rom)
+            return self._generate_bootloader_build(**self.get_all_active_memories(*self.get_region_list()))
```
Given that we only pass in the dict, and we don't have any other arguments, let's drop all of the `*` here.

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -696,7 +734,8 @@ def _assign_new_offset(rom_start, start, new_offset, region_name):
         return newstart
 
-    def _generate_bootloader_build(self, rom_start, rom_size):
+    def _generate_bootloader_build(self, **active_rom_regions):
```
How about we drop `**`, as it adds no value, and we do regions -> memories, and we drop the word "active", as it's not adding anything. 
```suggestion
    def _generate_bootloader_build(self, rom_memories):
```

bridadan commented on tools/toolchains/__init__.py:
```diff
@@ -732,13 +732,29 @@ def add_regions(self):
+
+        try:
+            # Add all available ROM regions to build profile
+            rom_available_regions = self.config.get_all_active_memories(*self.config.get_region_list())
```
Going with @theotherjimmy's suggestions, you'll want to drop the `*` here.

bridadan commented on tools/toolchains/__init__.py:
```diff
@@ -732,13 +732,29 @@ def add_regions(self):
-                suffixes=["_START", "_SIZE"]
-            )
+            # Add all available RAM regions to build profile
+            ram_available_regions = self.config.get_all_active_memories(*self.config.get_region_list('RAM'))
```
Going with @theotherjimmy's suggestions, you'll want to drop the `*` here.

bridadan commented on tools/config/__init__.py:
```diff
@@ -780,7 +819,8 @@ def report(self):
                 'library_configs': map(relpath, self.processed_configs.keys())}
 
-    def _generate_linker_overrides(self, rom_start, rom_size):
+    def _generate_linker_overrides(self, **active_rom_regions):
```
Same suggestion as @theotherjimmy:
```suggestion
    def _generate_linker_overrides(self, rom_regions):
```

bridadan commented on tools/config/__init__.py:
```diff
@@ -610,24 +610,62 @@ def _get_mem_specs(self, memories, cmsis_part, exception_text):
+                "project"
+            )
+            if region=='IROM1' or region=='PROGRAM_FLASH':
+                
```
Nit pick: Would you mind removing this new line? It isn't consistent with the other `elif`/`else` statements.
```suggestion
```

bridadan reviewed (changes requested):
Definitely looking better! Just echoing some of @theotherjimmy's comments.

Commit:
Add RAM/ROM memory statistics to system stats structure

Internal RAM / ROM memory size and start address of target can be
fetched using `mbed_stats_sys_get()` API.

deepikabhavnani commented:
@aashishc1988  - ~~I tried this PR for K64F to fetch RAM/ROM sizes and it looks RAM size is incorrect.~~

Steps to reproduce: mbed compile -m K64F -t GCC_ARM -v

As per the build log, defines passed are:
` -DMBED_ROM_START=0x0 -DMBED_ROM_SIZE=0x100000 -DMBED_RAM_START=0x20000000 -**DMBED_RAM_SIZE=0x30000** -DMBED_RAM1_START=0x1fff0000 -DMBED_RAM1_SIZE=0x10000 
`
As per CMSIS pack : `"SRAM_UPPER": {"start": "0x20000000", "size": "0x020000"}}`
Looks like DMBED_RAM_SIZE - is addition of SRAM_Upper and SRAM_lower sizes

Verified it again in `/tools/arm_pack_manager/index.json` and SRAM size is 0x30000 for MK64FN1M0xxx12

aashishc1988 commented:
Adding in the suggested changes.

bridadan reviewed (approved):
Nice work!

Commit:
Add tests to verify RAM/ROM sizes

0xc0170 commented:
@deepikabhavnani Waiting for your review (you have one PR depending on this one), please review

deepikabhavnani reviewed (changes requested):
Please squash the commits and add commit message describing the change

deepikabhavnani commented:
@0xc0170 - I have tested this PR with dependent PR for single target, so it should be good to go after appropriate commit messages are added

@cmonr - PR adds few defines to build steps, which can be verified by https://github.com/ARMmbed/mbed-os/pull/8661. Can we add both of them in single CI test as 8661 is essentially a test for this PR

aashishc1988 committed:
Fetch RAM/ROM information from CMSIS pack and add as defines

RAM/ROM memory data of target is required for statistics and linker
files, goal here is to fetch maximum 4 RAM/ROM regions from CMSIS pack
and make them available for C/C++/Linker flags as defines

cmonr commented:
**Info**: This PR is being tested as a part of a rollup PR (#8669).

cmonr commented:
@aashishc1988 Please refer to the following build error: https://github.com/ARMmbed/mbed-os/pull/8669#issuecomment-436758370

aashishc1988 committed:
Fixing some corner cases
1. Adding SRAM in available ram
2. If the target doesn't exist in cmsis but targets.json, handle accordingly and raise apt exceptions where needed.
3. If no sram is provided, raise exception

aashishc1988 commented on tools/config/__init__.py:
```diff
@@ -610,24 +616,72 @@ def _get_mem_specs(self, memories, cmsis_part, exception_text):
+        # Find which memory we are dealing with, RAM/ROM
+        active_memory = 'RAM' if any('RAM' in mem_list for mem_list in memory_list) else 'ROM'
+        
+        try:
```
Added this try/catch exception to handler cases where target doesnt exist in cmsis but in targets.json with ram and rom start/size defined

aashishc1988 commented on tools/config/__init__.py:
```diff
@@ -610,24 +616,72 @@ def _get_mem_specs(self, memories, cmsis_part, exception_text):
+
+        """For targets that do not have ram/rom defined, create dict with
+        start/size as 0 """
+        if not valid_memories:
```
Another corner case

aashishc1988 commented:
Added a new commit to handle some corner cases. I have added comments on those sections just so it is easier to do a comparison on what changed from last commit.

aashishc1988 commented on tools/config/__init__.py:
```diff
@@ -71,6 +71,14 @@
+# List of all possible ram memories that can be available for a target
+RAM_ALL_MEMORIES = ['IRAM1', 'IRAM2', 'IRAM3', 'IRAM4', 'SRAM_OC', \
+                    'SRAM_ITC', 'SRAM_DTC', 'SRAM_UPPER', 'SRAM_LOWER', \
+                    'SRAM']
```
Added SRAM

aashishc1988 commented on tools/config/__init__.py:
```diff
@@ -588,8 +596,6 @@ def sectors(self):
 
     def _get_cmsis_part(self):
-        if not getattr(self.target, "bootloader_supported", False):
-            raise ConfigException("Bootloader not supported on this target.")
```
Got rid of this and added this to regions property

aashishc1988 commented:
@theotherjimmy @bridadan @deepikabhavnani 
Please review

bridadan commented on tools/config/__init__.py:
```diff
@@ -632,10 +631,32 @@ def get_all_active_memories(self, memory_list):
+        """For targets that do not have ram/rom defined, create dict with
+        start/size as 0 """
+        if not valid_memories:
+            available_memories[active_memory] = [0, 0]
```
Is this how we currently handle this case? Or would raising a `ConfigException` be more appropriate here?

theotherjimmy replied:
This is not how we currently handle this case. Please raise the ConfigException like we had done before.

aashishc1988 replied:
But thats not how we were handling it before. I think my comment is misleading here. This part of code is to handle different memories that exists under RAM/ROM, for example, for STM32L443RC, we do have IROM defined, but we do not have any ram defined. In this case, we still need to create defines, MBED_RAM_START/SIZE, unless we intend to ifdef that in the C code, which is fine by me.

jeromecoutant replied:
Hi

> for STM32L443RC, we do have IROM defined

Issue is known since #5968 
@theotherjimmy

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -610,24 +616,72 @@ def _get_mem_specs(self, memories, cmsis_part, exception_text):
+            with ram and rom start/size defined"""
+            if hasattr(self.target, "mbed_ram_start") and \
+               hasattr(self.target, "mbed_rom_start"):
+               mem_start = getattr(self.target, "mbed_" + active_memory.lower() + "_start")
```
This should be indented 4 spaces (currently 3), and not match the prior line.

aashishc1988 commented on tools/toolchains/__init__.py:
```diff
@@ -693,14 +693,19 @@ def mem_stats(self, map):
+                ld_string = self.make_ld_define(*ld_string)
+                self.ld.append(ld_string)
+                self.flags["ld"].append(ld_string)
+    
```
Adding linker defines where needed.

aashishc1988 commented:
There was some misunderstanding where I assumed that if ram/rom memory sections are not present, we treat defines as zero, but that is not true. Removed that part of the code, and now making sure linker flags are properly created.

cmonr commented:
@aashishc1988 @deepikabhavnani Fine with doing test builds to iron out issues, especially since CI is just about empty right now, but would prefer to have the Travis CI smoke tests passing first.

deepikabhavnani commented:
@bridadan @theotherjimmy - This is now ready for review

aashishc1988 committed:
Missed one change to convert into right data type

0xc0170 commented:
/morph build

bridadan reviewed (approved):
Really nice work!

deepikabhavnani commented:
@ARMmbed/mbed-os-maintainers  - This PR will be tested with https://github.com/ARMmbed/mbed-os/pull/8661. And we expect some build failures, since few devices do not have RAM/ROM information in CMSIS packs, which needs to be fixed.

Please consider building both of them together for CI testing

cmonr commented:
Will do once CI has stabilized.

adbridge commented:
> 
> 
> @ARMmbed/mbed-os-maintainers - This PR will be tested with #8661. And we expect some build failures, since few devices do not have RAM/ROM information in CMSIS packs, which needs to be fixed.
> 
> Please consider building both of them together for CI testing

@deepikabhavnani This is not how the CI is designed to work! Could you please combine these two PRs together if neither will work without the other!

deepikabhavnani commented:
Added commits from https://github.com/ARMmbed/mbed-os/pull/8661.

cmonr commented:
/morph build

cmonr commented:
My bad. Accidentally stopped the build because I thought the reference was a new commit.

However, the build issues in the `jenkins-ci` checks should be fixed before we restart CI.

cmonr commented:
Something appears wrong with the REALTEK_RTL8195AM target with this PR.

Unfortunately, that's all I'm able to glean from the job itself.

@studavekar Would you happen to know how to access those build logs? Best I can see is the command being run, but no idea how the job is failing.

aashishc1988 committed:
The check for managed bootloader support should be in regions and we need a try/catch when we call it.

Moved "bootloader_not_supported" check to where it was and handle that exception at only one place.

Removed ram/rom size info for realtek from targets.json. THe info we have is not correct.

was Not handling config exceptions from regions and ram_regions property

adding rom-ram info for REALTEK_RTL8195AM

aashishc1988 commented:
@cmonr Latest commit should fix REALTEK_RTL8195AM failures

studavekar commented:
> @studavekar Would you happen to know how to access those build logs? Best I can see is the command being run, but no idea how the job is failing.

@OPpuolitaival should have some insight on that.

deepikabhavnani commented:
Resolved travis failures, good to go for CI

aashishc1988 committed:
ifdef ram/rom size/start in case they dont exist

cmonr commented:
/morph build

0xc0170 commented:
Please review the test failures (one target , all toolchains)

deepikabhavnani commented:
> Please review the test failures (one target , all toolchains)

Its fixed now

aashishc1988 commented:
It seems the checkout itself was failed. @cmonr

cmonr commented:
@aashishc1988 Yup. Job was accidentally triggered and is in a weird state. Looking into ways to mitigate. No PRs action needed.

cmonr commented:
/morph build

deepikabhavnani commented:
NRF error looks related to UART, seems I have seen this before and was resolved?

`[1542407234.13][CONN][RXD] :144::FAIL: Expected 'passed' Was 'pased'`

deepikabhavnani commented:
Not able to find failures in jenkins-ci, logs and results say success.. :-(

cmonr commented:
> Not able to find failures in jenkins-ci, logs and results say success.. :-(

NRF issue. Will restart in a bit.

Retriggered the pr-head job. Looks odd.

0xc0170 commented:
Info: This PR has been re-bundled into a new rollup PR (#8800).

No further work is needed here, as once that PR is merged, this PR will also be closed and marked as merged.
If any more commits are made in this PR, this PR will remain open and have to go through CI on its own.

cmonr commented:
Restarting CI.
